### PR TITLE
Convert a few more locally-owned loops with filters.

### PR DESCRIPTION
**Pull request by bangerth**
Another follow-up to #12952.

/rebuild

### CODE REVIEW

File: source/distributed/shared_tria.cc [hunk 1]
```diff
@@ -283,13 +283,12 @@ namespace parallel
...
-          cell = this->begin_active(),
-          endc = this->end();
-        for (; cell != endc; ++cell)
-          if (cell->is_locally_owned())
-            n_my_cells += 1;
+        for (const auto &cell : this->active_cell_iterators() |
+                                  IteratorFilters::LocallyOwnedCell())
+          {
+            (void)cell;
+            ++n_my_cells;
```
Comment (bangerth) [inline source/distributed/shared_tria.cc:290]: It is disappointing that one has to give the loop variable a name, but that appears to be the case based on the standard.

Comment (marcfehling) [reply source/distributed/shared_tria.cc:290]: Would it be more elegant to use `std::count_if` here?

File: source/distributed/tria_base.cc [hunk 2]
```diff
@@ -178,9 +179,12 @@ namespace parallel
     }
 
     if (this->n_levels() > 0)
-      for (const auto &cell : this->active_cell_iterators())
-        if (cell->subdomain_id() == my_subdomain)
+      for (const auto &cell : this->active_cell_iterators() |
+                                IteratorFilters::SubdomainEqualTo(my_subdomain))
+        {
+          (void)cell;
```
Comment (marcfehling) [inline source/distributed/tria_base.cc:185]: see above

File: source/distributed/tria.cc [hunk 3]
```diff
@@ -22,6 +22,7 @@
 #include <deal.II/distributed/p4est_wrappers.h>
 #include <deal.II/distributed/tria.h>
 
+#include <deal.II/grid/filtered_iterator.h>
```
Comment (marcfehling) [inline source/distributed/tria.cc:25]: Doesn't look like you need it here
```suggestion
```

File: source/distributed/shared_tria.cc [hunk 4]
```diff
@@ -302,12 +301,12 @@ namespace parallel
...
-          typename parallel::shared::Triangulation<dim, spacedim>::cell_iterator
-            cell = this->begin(),
-            endc = this->end();
-          for (; cell != endc; ++cell)
-            if (cell->is_locally_owned_on_level())
-              n_my_cells += 1;
+          for (const auto &cell : this->cell_iterators() |
+                                    IteratorFilters::LocallyOwnedLevelCell())
+            {
+              (void)cell;
```
Comment (marcfehling) [inline source/distributed/shared_tria.cc:307]: see above

File: source/distributed/shared_tria.cc (hunk 1 above)
Comment (marcfehling) [reply source/distributed/shared_tria.cc:290]: Or use what you did above in `source/distributed/repartitioning_policy_tools.cc`.
```
const auto locally_owned_cells =
    tria_in.active_cell_iterators() | IteratorFilters::LocallyOwnedCell();
const unsigned int n_locally_owned_active_cells =
    std::count(locally_owned_cells.begin(), locally_owned_cells.end());
```

File: source/grid/grid_tools_cache.cc [hunk 5]
```diff
@@ -156,10 +156,10 @@ namespace GridTools
           typename Triangulation<dim, spacedim>::active_cell_iterator>>
           boxes;
         boxes.reserve(tria->n_active_cells());
-        for (const auto &cell : tria->active_cell_iterators())
+        for (const auto &cell : tria->active_cell_iterators() |
+                                  IteratorFilters::LocallyOwnedCell())
           if (cell->is_locally_owned())
```
Comment (marcfehling) [inline source/grid/grid_tools_cache.cc:161]: You don't need this anymore I guess.
```suggestion
```

**Review by marcfehling — CHANGES_REQUESTED** [review]

**Comment by bangerth** [comment]
Ah yes, `std::count_if` is a clever solution. Thanks for pointing that out!

File: source/distributed/tria_base.cc [hunk 6]
```diff
@@ -1518,8 +1526,8 @@ namespace parallel
           const MPI_Offset my_global_file_position =
             static_cast<MPI_Offset>(global_first_cell) * sizeof(unsigned int);
 
-          // It is very unlikely that a single process has more than 2 billion
-          // cells, but we might as well check.
+          // It is very unlikely that a single process has more than 2
+          // billion cells, but we might as well check.
```
Comment (jppelteret) [inline source/distributed/tria_base.cc:1530]: If you're going to push "billon" onto the next line, then maybe it makes sense to move the number "2" as well.

**Review by jppelteret — APPROVED** [review]
This is pretty cool :-)

**Comment by bangerth** [comment]
Adjusted as requested!

**Comment by bangerth** [comment]
Seems to work now!

@marcfehling You put the block on this PR. Want to take another look?

**Comment by bangerth** [comment]
Ping?

**Review by masterleinad — APPROVED** [review]

File: source/distributed/repartitioning_policy_tools.cc [hunk 7]
```diff
@@ -206,10 +206,10 @@ namespace RepartitioningPolicyTools
...
     // step 1) check if all processes have enough cells
 
-    unsigned int n_locally_owned_active_cells = 0;
-    for (const auto &cell : tria_in.active_cell_iterators())
-      if (cell->is_locally_owned())
-        ++n_locally_owned_active_cells;
+    const auto locally_owned_cells =
+      tria_in.active_cell_iterators() | IteratorFilters::LocallyOwnedCell();
+    const unsigned int n_locally_owned_active_cells =
+      std::distance(locally_owned_cells.begin(), locally_owned_cells.end());
```
Comment (marcfehling) [inline source/distributed/repartitioning_policy_tools.cc:212]: If you want, you can use `std::count_if` here, too.

**Review by marcfehling — APPROVED** [review]
Sorry for the late feedback. Forgot about this open PR over the holidays.

File: source/distributed/repartitioning_policy_tools.cc (hunk 7 above)
Comment (bangerth) [reply source/distributed/repartitioning_policy_tools.cc:212]: Yes, so done!

File: source/distributed/repartitioning_policy_tools.cc [hunk 8]
```diff
@@ -206,10 +206,10 @@ namespace RepartitioningPolicyTools
 
     // step 1) check if all processes have enough cells
 
-    unsigned int n_locally_owned_active_cells = 0;
-    for (const auto &cell : tria_in.active_cell_iterators())
-      if (cell->is_locally_owned())
-        ++n_locally_owned_active_cells;
+    const unsigned int n_locally_owned_active_cells =
+      std::count_if(tria_in.begin_active(),
+                    tria_in.end(),
```
Comment (marcfehling) [inline source/distributed/repartitioning_policy_tools.cc:211]:
```suggestion
                    typename Triangulation<dim, spacedim>::active_cell_iterator(
                      tria_in.end()),
```
This should fix the compiler problem
```
/home/runner/work/dealii/dealii/source/distributed/repartitioning_policy_tools.cc:210:20: note:   deduced conflicting types for parameter ‘_IIter’ (‘dealii::TriaActiveIterator<dealii::CellAccessor<1, 1> >’ and ‘dealii::TriaIterator<dealii::CellAccessor<1, 1> >’)
```

**Review by marcfehling — APPROVED** [review]

**Review by marcfehling — APPROVED** [review]

**Pull request merged**